liquidctl/keyval.py

```diff
@@ -87,30 +97,112 @@ def store(self, key, value):
+
         _LOGGER.debug('stored %s=%r (in %s)', key, value, path)
 
+    def load_store(self, key, func):
+        value = None
+        new_value = None
+
+        path = os.path.join(self._write_dir, key)
+
+        # lock the destination as soon as possible
+        with self._open_with_lock(path, os.O_RDWR | os.O_CREAT) as f:
+
+            # still traverse all possible locations to find the current value
+            for base in self._read_dirs:
+                read_path = os.path.join(base, key)
+                if not os.path.isfile(read_path):
+                    continue
+                try:
+                    if os.path.samefile(read_path, path):
+                        # we already have an exclusive lock to this file
+                        data = f.read().strip()
+                        f.seek(0)
+                    else:
+                        with self._open_with_lock(read_path, os.O_RDWR, shared=True) as aux:
+                            data = aux.read().strip()
+
+                    if not data:
+                        continue
+
+                    value = literal_eval(data)
+                    _LOGGER.debug('loaded %s=%r (from %s)', key, value, read_path)
+                except OSError as err:
+                    _LOGGER.warning('%s exists but could not be read: %s', read_path, err)
+                except ValueError as err:
+                    _LOGGER.warning('%s exists but was corrupted: %s', key, err)
+                else:
+                    break
+            else:
+                _LOGGER.debug('no data (file) found for %s', key)
+
+            new_value = func(value)
+
+            data = repr(new_value)
+            assert literal_eval(data) == new_value, 'encode/decode roundtrip fails'
+            f.write(data)
+            f.truncate()
+            f.flush()  # ensure flushing before automatic unlocking
+
+            _LOGGER.debug('replaced with %s=%r (stored in %s)', key, new_value, path)
+
+        return (value, new_value)
+
+    @contextmanager
+    def _open_with_lock(self, path, flags, *, shared=False):
+        with os.fdopen(os.open(path, flags), mode='r+') as f:
+            if sys.platform == 'win32':
+                msvcrt.locking(f.fileno(), msvcrt.LK_LOCK, 1)
+            elif shared:
+                fcntl.flock(f, fcntl.LOCK_SH)
+            else:
+                fcntl.flock(f, fcntl.LOCK_EX)
+
+            try:
+                yield f
+            finally:
+                if sys.platform == 'win32':
+                    f.seek(0)
+                    msvcrt.locking(f.fileno(), msvcrt.LK_UNLCK, 1)
+                else:
+                    fcntl.flock(f, fcntl.LOCK_UN)
+
 
 class RuntimeStorage:
     """Unstable API."""
 
-    def __init__(self, key_prefixes):
-        self._backend = _FilesystemBackend(key_prefixes)
+    def __init__(self, key_prefixes, backend=None):
+        self._backend = backend if backend is not None else _FilesystemBackend(key_prefixes)
```

Comment:
Self nitpick: this could be written in a more clear way.

Also, should be accept `runtime_dirs`  instead of requiring the tests to construct a private `_FilesystemBackend` class?  This would also avoid having the tests pass `key_prefixes` both to the backend and to the `RuntimeStorage` wrapper.

Reply:
imo passing a backend to the test could make sense because it would allow us to swap it out with a mock to speed up the driver tests instead of writing to the disk, (which from what my understanding was is best practice anyway?)

Reply:
We used to mock the backend without any actual FS access, but in this branch I actually made all tests revert back to a real FS backend, albeit with temporary and isolated locations.

Using a mocked backend requires keeping it in sync with the real implementation, something we weren't great at.  And that extra work and risk for a small performance improvement in our test suite is just not worth it, IMO.